tests/returntypes/test_parameter_shift_hessian_new.py

```diff
@@ -0,0 +1,1849 @@
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(x, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.Hadamard(1))
+
+        expected = (-np.cos(x), -np.cos(x) / np.sqrt(2))
+
+        tapes, fn = qml.gradients.param_shift_hessian(tape)
+        hessian = fn(qml.execute(tapes, dev, gradient_fn=None))
+
+        assert isinstance(hessian, tuple)
+        assert len(hessian) == 2
+
+        for hess, exp in zip(hessian, expected):
+            assert isinstance(hess, np.ndarray)
+            assert hess.shape == ()
+            assert np.allclose(hess, exp)
+
+    def test_multi_expval_probs(self):
+        """Test that the correct hessian is calculated for a tape with single RX operator
+        and both expval and probability outputs"""
+        dev = qml.device("default.qubit", wires=2)
+
+        x = np.array(0.1, requires_grad=True)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(x, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.probs(wires=[0, 1])
+
+        expected = (-np.cos(x), 0.5 * np.cos(x) * np.array([-1, 0, 0, 1]))
+
+        tapes, fn = qml.gradients.param_shift_hessian(tape)
+        hessian = fn(qml.execute(tapes, dev, gradient_fn=None))
+
+        assert isinstance(hessian, tuple)
+        assert len(hessian) == 2
+
+        for hess, exp in zip(hessian, expected):
+            assert isinstance(hess, np.ndarray)
+            assert hess.shape == exp.shape
+            assert np.allclose(hess, exp)
+
+    def test_multi_probs(self):
+        """Test that the correct hessian is calculated for a tape with single RX operator
+        and multiple probability outputs"""
+        dev = qml.device("default.qubit", wires=2)
+
+        x = np.array(0.1, requires_grad=True)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(x, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.probs(wires=[0])
+            qml.probs(wires=[0, 1])
+
+        expected = (0.5 * np.cos(x) * np.array([-1, 1]), 0.5 * np.cos(x) * np.array([-1, 0, 0, 1]))
+
+        tapes, fn = qml.gradients.param_shift_hessian(tape)
+        hessian = fn(qml.execute(tapes, dev, gradient_fn=None))
+
+        assert isinstance(hessian, tuple)
+        assert len(hessian) == 2
+
+        for hess, exp in zip(hessian, expected):
+            assert isinstance(hess, np.ndarray)
+            assert hess.shape == exp.shape
+            assert np.allclose(hess, exp)
+
+    def test_single_expval_multi_params(self):
+        """Test that the correct hessian is calculated for a tape with multiple operators
+        and single expectation value output"""
+        dev = qml.device("default.qubit", wires=2)
+
+        x = np.array([0.1, 0.4], requires_grad=True)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(x[0], wires=0)
+            qml.RY(x[1], wires=1)
+            qml.CNOT(wires=[0, 1])
```

Comment:
Minor: does it add value to have CNOTs in cases where they don't affect the Hessian? (No clear preference here; it may cause confusion for some readers, but it also serves as a small test step verifying that a CNOT indeed does not influence the value of the Hessian.)

Reply:
I originally put CNOT's in all of the tests to make the gradient computation by hand easier (since the density matrix becomes diagonal). Personally I'd prefer to leave them in.